feat: external surveys

lucasheriques opened the pull request:
## Problem

First part necessary for https://github.com/PostHog/posthog/issues/21071
Closes https://github.com/PostHog/posthog/issues/27380
Don't merge before https://github.com/PostHog/posthog-js/pull/2091

Right now, we don't have any other way to answer a survey besides via in-app surveys. This will change with this feature, so we also need to update it on a lot of places like our website and docs.

## Changes

Adds a new route, `/external_surveys/:survey_id` that's rendering HTML to display a survey. It uses posthog-js script for that, configured with the client's settings. [Check an example live here](https://us.posthog.com/external_surveys/01984280-fc8a-0000-28a5-01078e2d553f)

![CleanShot 2025-07-08 at 02 50 29@2x](https://github.com/user-attachments/assets/dcad1768-afce-4288-918f-3e54a8448b53)

It only works for surveys which are marked as external surveys, which is done in the Presentation tab:

<img width="1798" height="526" alt="image" src="https://github.com/user-attachments/assets/3aca1fa5-115f-401e-bb1a-207629b76a34" />


Once done, you can find out the survey link via the survey overview:

![CleanShot 2025-07-08 at 02 51 02@2x](https://github.com/user-attachments/assets/5332f103-93d5-4499-988f-42d520de2ffe)

Error page:

![CleanShot 2025-07-14 at 21 30 07@2x](https://github.com/user-attachments/assets/eb8ec717-1e3c-4a44-848d-60783a6e8046)

## Did you write or update any docs for this change?

<!-- Engineers are responsible for doing the first pass at documenting their features and/or code.  -->
Docs will be done after this PR. For now, I'm releasing it under a feature flag so it'll not be seen by any customers.

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

lucasheriques committed:
feat: fullscreen survey mvp

lucasheriques committed:
ssr the public survey

marandaneto commented:
once this is out, we have to fix docs and a few things on the product page such as:

> PostHog offers multi-step surveys, but they won't be full-page forms such as Typeform or Google Forms

found here -> https://posthog.com/surveys

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
remove survey scene

lucasheriques committed:
remove public survey

lucasheriques committed:
allow collecting answers publicly

github-actions[bot] committed:
Update UI snapshots for `chromium` (13)

github-actions[bot] committed:
Update UI snapshots for `chromium` (10)

lucasheriques committed:
new styles

lucasheriques committed:
dont apply font family for external surveys

lucasheriques committed:
potential last changes

lucasheriques committed:
last style change

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
more style changes

lucasheriques committed:
add test to copy the survey link

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

lucasheriques committed:
posthog-like error page and tests

lucasheriques commented on posthog/templates/surveys/public_survey.html:
```diff
@@ -0,0 +1,948 @@
+        });
+
+        // Fallback timeout in case onSurveysLoaded doesn't fire
+        setTimeout(() => {
+            if (!surveyRendered) {
+                renderSurveyWithRetry();
+            }
+        }, 5000);
+
```
i removed, this isn't needed

lucasheriques committed:
review changes

lucasheriques committed:
copy

lucasheriques committed:
fix tests

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

github-actions[bot] committed:
Update UI snapshots for `chromium` (6)

lucasheriques committed:
fix test

lucasheriques commented:
@ioannisj do you think it might be better to instead use a whole separate type then, at least for the first release? it'll require some changes but not too much

marandaneto commented on frontend/src/lib/constants.tsx:
```diff
@@ -170,8 +170,7 @@ export const FEATURE_FLAGS = {
     SURVEYS_ADAPTIVE_LIMITS: 'surveys-adaptive-limits', // owner: #team-surveys
     SURVEYS_ACTIONS: 'surveys-actions', // owner: #team-surveys
     SURVEYS_CUSTOM_FONTS: 'surveys-custom-fonts', // owner: #team-surveys
-    SURVEYS_PARTIAL_RESPONSES: 'surveys-partial-responses', // owner: #team-surveys
-    SURVEYS_NEW_QUESTION_VIZ: 'surveys-new-question-viz', // owner: #team-surveys
```
intentional removal?

lucasheriques replied:
yeah not using those anymore

marandaneto commented on frontend/src/scenes/surveys/CopySurveyLink.tsx:
```diff
@@ -0,0 +1,18 @@
+        <LemonButton
+            icon={<IconLink />}
+            onClick={() => {
+                const url = new URL(window.location.origin)
```
Will that address the eu.* or us.*?

marandaneto commented on frontend/src/scenes/surveys/SurveyStatsSummary.tsx:
```diff
@@ -159,24 +160,30 @@ function SurveyStatsStackedBar({
-                )}
+                <div className="flex items-center gap-2">
+                    {isPubliclyShareable && <CopySurveyLink surveyId={survey.id} />}
+                    <LemonSwitch
+                        checked={isPubliclyShareable}
+                        onChange={(checked) => updateSurvey({ is_publicly_shareable: checked })}
+                        label="Collect responses via external link"
+                    />
```
will this be visible for every presentation type? eg popver, widget and API? should that be the case?

lucasheriques replied:
talked more about it in the discussions below, but will change so it instead is its own type

marandaneto commented on posthog/api/survey.py:
```diff
@@ -1386,9 +1394,178 @@ def surveys(request: Request):
+            request,
+            "surveys/error.html",
+            {
+                "error_title": "Invalid Request",
```
```suggestion
                "error_title": "Invalid request",
```

marandaneto commented on posthog/api/survey.py:
```diff
@@ -1386,9 +1394,178 @@ def surveys(request: Request):
+
+    response = render(request, "surveys/public_survey.html", context)
+
+    # Security headers
+    response["X-Frame-Options"] = "DENY"
+    response["X-Content-Type-Options"] = "nosniff"
+    response["Referrer-Policy"] = "strict-origin-when-cross-origin"
+    response["Permissions-Policy"] = "accelerometer=(), camera=(), microphone=(), geolocation=()"
+    response["X-XSS-Protection"] = "1; mode=block"
+
+    # Cache headers
+    response["Cache-Control"] = f"public, max-age={CACHE_TIMEOUT_SECONDS}"
+    response["Vary"] = "Accept-Encoding"  # Enable compression caching
```
do we need all of that? do we set this elsewhere or how did we come up with those headers? just curious since i dont know much about cors, etc

lucasheriques commented on frontend/src/scenes/surveys/CopySurveyLink.tsx:
```diff
@@ -0,0 +1,18 @@
+            icon={<IconLink />}
+            onClick={() => {
+                const url = new URL(window.location.origin)
+                url.pathname = `/surveys/${surveyId}`
```
we still capture them! because we're using the `renderSurvey` method from the SDK, which emits those events, same as other surveys

marandaneto replied:
will those events be captured within our api key or their api key since its now hosted by us? will they be able to analyze their own surveys stats?

lucasheriques replied:
it's captured with their own API key. From the survey id, we also fetch the team, and we include that in the payload to render the HTML and init the SDK with the owner team data. all events are captured just as if they were on their own app

lucasheriques committed:
use UUIDT instead

lucasheriques commented on posthog/api/survey.py:
```diff
@@ -1386,9 +1394,178 @@ def surveys(request: Request):
+        return cors_response(request, HttpResponse(""))
+
+    # Input validation
+    if not is_valid_uuid(survey_id) or len(survey_id) > SURVEY_ID_MAX_LENGTH:
```
removed

lucasheriques committed:
casing

lucasheriques committed:
remove other headers

lucasheriques commented on posthog/templates/surveys/error.html:
```diff
@@ -0,0 +1,266 @@
+<!DOCTYPE html>
```
![CleanShot 2025-07-14 at 21 30 07@2x](https://github.com/user-attachments/assets/eb8ec717-1e3c-4a44-848d-60783a6e8046)

also added to the PR

lucasheriques commented on posthog/templates/surveys/public_survey.html:
```diff
@@ -0,0 +1,935 @@
+            applySurveyAppearance(window.surveyAppearance);
+        }
+
+        // Load PostHog from CDN
```
actually this is not correct. We're not loading the usual posthog React app. We're just rendering a HTML page (this one you commented). So we need to load the snippet here regardless, since we also need to init it according to the customer's config

lucasheriques commented:
> A link I will be sending to folks via email or some other way
All the display conditions won't apply here since folks will have the link anyway

correct

> If I have a link, will the complete conditions apply in the API as well? what will happen?

Only partial responses and `How long would you like to collect survey responses?` are relevant here. How often we should show the survey is irrelevant.

> What if I want a survey specifically for emails, and suddenly now it's a popover in my app as well - I will get confusing/mixed results
It isn't very clear for me, being a popover type with a shareable link, which is not a popover anymore and i think this applies to all presentation types?

Talked more about it above, but I agree with the thoughts: will make the external/hosted survey its own type.

the shareable link is only for anonymous users since there's no way to find out who's answering the survey, is that intentional?

For now, intentional, but before releasing publicly I'll add a way to identifying responses. Probably through URL parameters for the MVP as you mentioned on Slack too

marandaneto commented on posthog/api/survey.py:
```diff
@@ -1386,9 +1394,178 @@ def surveys(request: Request):
+
+@csrf_exempt
+@axes_dispatch
+def public_survey_page(request, survey_id: str):
```
is the current API and approach extendable for identified users in the near feature? or will we need to break links?

lucasheriques replied:
i just tested it, for now it's a very simple thing: just add a query parameter, `distinct_id`, [and use that for identifying users](https://github.com/PostHog/posthog/pull/33948/files#diff-fdea57bef4ea1f195c8313118abe5b886b35d55386ec5d1d57abee11d1c9a5a2R903-R906). just tested and it worked fine 🙏 

though in the future we might extend this by instead asking the user to provide something like their email directly.

No need for breaking links, as url will forever be `external_surveys`

marandaneto commented:
> For now, intentional, but before releasing publicly I'll add a way to identify responses. Probably through URL parameters for the MVP as you mentioned on Slack too

@lucasheriques since this is different than all the other types, this should be in the UI as well before people start using and expecting to know which users have answered the survey, either as a label or tooltip or something right in their face to avoid potential issues/bunch of tickets

marandaneto commented on posthog/api/survey.py:
```diff
@@ -1356,7 +1365,7 @@ def get_surveys_response(team: Team):
 
 @csrf_exempt
-def surveys(request: Request):
+def surveys(request: Request, survey_id: str | None = None):
```
Should we add a new endpoint instead to separate things and be better able to estimate costs in the future? (or even better, scale control whenever something happens)
like /external_survey/$surveyId?distinctId=123&...

marandaneto replied:
or, would we be able to estimate costs for requests with and w/o `survey_id`?

lucasheriques replied:
now it's on a new endpoint, /external_surveys

lucasheriques committed:
merge master

lucasheriques committed:
use new type for external surveys

lucasheriques committed:
beta tag

lucasheriques committed:
change description

lucasheriques committed:
update api

lucasheriques committed:
why was this deleted?

lucasheriques committed:
this is not needed

lucasheriques committed:
this is not needed too

lucasheriques committed:
add tests and change URL

lucasheriques commented:
@marandaneto @ioannisj I updated so the URL is now `/external_surveys/:id` and also added tests for it. 

new type too:

![CleanShot 2025-07-15 at 17 23 14@2x](https://github.com/user-attachments/assets/97d38c59-59c7-40e3-96bb-653f0feb9fc8)

please check it out!

lucasheriques committed:
update copy link

lucasheriques committed:
identify user

marandaneto commented:
do we need changes on the SDK side to not display `SurveyType.ExternalSurvey` surveys as in app?
if we do those changes, how will older SDK versions behave? how can we mitigate this?
This includes Mobile SDKs @ioannisj

lucasheriques commented:
@marandaneto @ioannisj [here's the SDK PR](https://github.com/PostHog/posthog-js/pull/2091) to not evaluate any external surveys. I also checked the iOS/RN SDKs, and from the looks of it, no changes are needed there. But let me know if not!

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
linter

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
migrations

lucasheriques committed:
remove is_publicly_shareable

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
migrations

lucasheriques committed:
Revert "migrations"

This reverts commit 7aef3af9d2d9269a2e8f5ebd8cf24bfcefc89808.

lucasheriques committed:
migrations with deprecated

lucasheriques committed:
unnecesary test

github-actions[bot] committed:
Update query snapshots

marandaneto commented on frontend/src/types.ts:
```diff
@@ -3077,7 +3077,6 @@ export interface Survey {
     response_sampling_limit?: number | null
     response_sampling_daily_limits?: string[] | null
     enable_partial_responses?: boolean | null
-    is_publicly_shareable?: boolean | null
```
we need a PR to remove this from the DB right? before merging this
added here: https://github.com/PostHog/posthog/pull/33994/files

marandaneto commented on posthog/models/surveys/survey.py:
```diff
@@ -222,10 +224,13 @@ class Meta:
-        null=True,
-        blank=True,
-        help_text="Allow this survey to be accessed via public URL (https://app.posthog.com/surveys/[survey_id]) without authentication",
+    # Use the survey_type instead. If it's external_survey, it's publicly shareable.
+    is_publicly_shareable = deprecate_field(
+        models.BooleanField(
+            null=True,
+            blank=True,
+            help_text="Allow this survey to be accessed via public URL (https://app.posthog.com/surveys/[survey_id]) without authentication",
+        ),
```
this field should be removed since its not used anymore?

marandaneto commented on posthog/migrations/0799_alter_survey_type.py:
```diff
@@ -0,0 +1,28 @@
+                choices=[
+                    ("popover", "popover"),
+                    ("widget", "widget"),
+                    ("button", "button"),
```
whats this type?

lucasheriques replied:
those were always there, though i don't think they're used for anything. but makes sense to remove them since we're not really using, so will do

marandaneto commented on posthog/migrations/0799_alter_survey_type.py:
```diff
@@ -0,0 +1,28 @@
+                    ("widget", "widget"),
+                    ("button", "button"),
+                    ("email", "email"),
+                    ("full_screen", "full screen"),
+                    ("external_survey", "external survey"),
```
what's the difference between full_screen and external_survey here?

lucasheriques replied:
so besides the three ones we already have (popover, api, feedback button), we also had button, email and full_screen. We don't use them anywhere, but I preferred to not touch them. do you think se should remove them?

marandaneto commented on posthog/templates/surveys/public_survey.html:
```diff
@@ -0,0 +1,939 @@
+
+        posthog.init(window.projectConfig.token, config);
+        const distinctId = new URLSearchParams(window.location.search).get('distinct_id');
+        if (distinctId) {
+            posthog.identify(distinctId);
+        }
```
Users will need to pay for this event, which might not be expected/intentional.
Can we actually just mimic the identify behaviour without capturing the identify event?
what i mean is, the survey events should contain the 'distinctId' but we dont necessary need to capture an identify event, an idea is to init the SDK with the `bootstrap` object and the given `distinctId`

marandaneto commented on posthog/templates/surveys/public_survey.html:
```diff
@@ -0,0 +1,939 @@
+        const config = {
+            api_host: window.projectConfig.api_host,
+            disable_surveys_automatic_display: true,
+            debug: true,
```
should be false

marandaneto commented on posthog/templates/surveys/public_survey.html:
```diff
@@ -0,0 +1,939 @@
+            posthog.identify(distinctId);
+        }
+
+        let surveyRendered = false;
+        let retryCount = 0;
+        const maxRetries = 5;
```
why do we need this retry logic?

lucasheriques replied:
we don't, really. Claude suggested it when I asked it to review my code for production readiness, but looking again I don't see a case where it's needed. removed

marandaneto commented on posthog/api/survey.py:
```diff
@@ -1389,6 +1396,116 @@ def surveys(request: Request):
+
+@csrf_exempt
+@axes_dispatch
+def public_survey_page(request, survey_id: str):
```
This should be a public link that can load and render a survey with given parameters, such as survey ID, distinct ID, and more.
This link is declared in the Django API file.
Do you know if this is an API per se? I want to explore whether we should have approached this differently, such as hosting a simple app (HTML/JS) that then calls our APIs to load the survey, etc.
Currently, everything is handled within the API file, but it simply renders the HTML based on specific conditions (errors or not).
It's not a blocker; I want to make sure the "architecture" here makes sense.
For example, does this method require setting content-type, etc, all of that because it's an API handler instead of a "public page"?
The external survey run by our support team does exactly what i mention here.
WDYT?

lucasheriques replied:
I think the current approach makes sense. It's similar how we also do [public dashboards like this one](https://us.posthog.com/shared/fEjuk1e61iQjYiRjODPDjpbsyhIfpg) - they're also server-side rendered HTML in the API module ([api/sharing.py](https://github.com/PostHog/posthog/blob/feat/fullscreen-survey/posthog/api/sharing.py#L250))

the main benefits:
- Fast: One request vs waterfall (validate survey → get team → fetch token (all done on the server) → render and just wait for posthog-js to load)
- Simple: No client state management, loading states, or extra JS
- Secure: Server-side validation prevents survey ID enumeration
- Consistent: Same pattern as existing shared dashboards/insights

a separate HTML+JS app would be slower and more complex - we'd still need the same server-side validation logic anyway, just exposed as separate API endpoints. we do that for the CSAT survey, however, the CSAT survey is also doing a unique UI for them to be able to render that in one go. The goal is for external surveys to replace it.

The "API file" thing is just organization - half of posthog/api/ returns HTML templates. Django handles content-type headers automatically.

only reason to change would be if external survey load becomes massive, so that we don't have to share resources with the API. but I don't think we should worry about that, unless we actually see most of the requests are coming for external surveys.

makes sense?

marandaneto replied:
The whole 'simple/secure' doesn't apply since a separate URL/app would still be the same.
Regarding speed, it's more about latency since the whole system is distributed anyway, but the current approach has the database closer to the hosting app, which is true. I agree with that.
I didn't know that we do the same for some other things.
Let's keep it as is, thank you.
Last question, how the URL will look like? can you give an example with localhost and the whole path segments?

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

lucasheriques commented on posthog/migrations/0799_alter_survey_type.py:
```diff
@@ -0,0 +1,28 @@
+        migrations.AlterField(
+            model_name="survey",
+            name="type",
+            field=models.CharField(
```
it was always typed like that, but now the choices are different

lucasheriques replied:
yeah the migration is needed when the field changes. `python manage.py makemigrations --check --dry-run` is run every time we push a PR, and it will detect a migration is missing if we alter a model (even if not adding a new column, just changing the choices).

`AlterField` is the correct one for that. it'll will change the current column, so we update the choices (adding external_survey, removing the others not used)

lucasheriques committed:
use bootstrap instead

lucasheriques committed:
wrong order

lucasheriques committed:
wrong worder

lucasheriques committed:
typo

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
remove migration

lucasheriques committed:
remove retry logic

lucasheriques commented on frontend/src/scenes/surveys/SurveyEdit.tsx:
```diff
@@ -218,7 +219,7 @@ function SurveyCompletionConditions(): JSX.Element {
                 </LemonField>
             )}
-            <SurveyRepeatSchedule />
+            {survey.type !== SurveyType.ExternalSurvey && <SurveyRepeatSchedule />}
```
hmm I don't think it's needed. The feature flag status doesn't really matter for this

lucasheriques commented on frontend/src/scenes/surveys/SurveyOverview.tsx:
```diff
@@ -105,6 +110,14 @@ export function SurveyOverview(): JSX.Element {
                     {survey.enable_partial_responses ? 'Enabled' : 'Disabled'}
                 </SurveyOption>
+                {featureFlags[FEATURE_FLAGS.EXTERNAL_SURVEYS] && (
+                    <SurveyOption label="Responses via external link">
```
![CleanShot 2025-07-24 at 03 09 03@2x](https://github.com/user-attachments/assets/01f5e159-937e-455e-b2c2-3f4b1e4b68d1)

maybe something like this could work? liked the idea of showing up even when FF is disabled

marandaneto replied:
i think the approach here is show or not show depending on feature flag status.
if you want to advertize something that is coming, we already have the early preview feature
i created for emails at least https://us.posthog.com/project/2/early_access_features/01958a32-b0ba-0000-2a31-d47d5ab4c9fe
since we are about to merge this and its under a feature flag, i suggest creating a new one for external survyes and let people enroll and test themselves

lucasheriques committed:
changes to feature flag

github-actions[bot] committed:
Update UI snapshots for `chromium` (11)

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
remove type

lucasheriques committed:
add migration

lucasheriques committed:
fix FE

lucasheriques committed:
Revert "add migration"

This reverts commit 25d37f55120a5b06cc786fd28d73f27f3265b794.

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
re-add migration after merge

marandaneto reviewed (approved):
left a last [suggestion here](https://github.com/PostHog/posthog/pull/33948#discussion_r2227669726) otherwise LGTM

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

lucasheriques committed:
update schema

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

lucasheriques committed:
Merge branch 'master' into feat/fullscreen-survey

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)
